nmostesting/suites/IS0701Test.py

```diff
@@ -21,3 +28,322 @@ class IS0701Test(GenericTest):
+                    self.sources[source_id][sub_path] = sub.json()
+        except json.JSONDecodeError:
+            raise NMOSTestException(test.FAIL("Non-JSON response returned from Events API"))
+        except ValueError:
+            raise NMOSTestException(test.FAIL("Invalid response returned from Events API"))
+
+    def test_01(self, test):
+        """Each Source state identity includes the correct ID"""
+        self.do_collect_sources(test)
+
+        if len(self.sources) == 0:
+            return test.UNCLEAR("No sources were returned from Events API")
+
+        try:
+            for source_id in self.sources:
+                if source_id != self.sources[source_id]["state"]["identity"]["source_id"]:
+                    return test.FAIL("Source {} state has incorrect source_id".format(source_id))
+        except KeyError:
+            return test.FAIL("Source {} JSON data is invalid".format(source_id))
+
+        return test.PASS()
+
+    def test_02(self, test):
+        """Each Source state identity does not include a Flow ID"""
+        self.do_collect_sources(test)
+
+        if len(self.sources) == 0:
+            return test.UNCLEAR("No sources were returned from Events API")
+
+        try:
+            for source_id in self.sources:
+                if "flow_id" in self.sources[source_id]["state"]["identity"]:
+                    return test.FAIL("Source {} state has flow_id which is not permitted".format(source_id))
+        except KeyError:
+            return test.FAIL("Source {} JSON data is invalid".format(source_id))
+
+        return test.PASS()
+
+    def test_03(self, test):
+        """Each Source type 'type' is a base type of the state 'event_type'"""
+        self.do_collect_sources(test)
+
+        if len(self.sources) == 0:
+            return test.UNCLEAR("No sources were returned from Events API")
+
+        try:
+            for source_id in self.sources:
+                base_type = self.sources[source_id]["state"]["event_type"].split("/")[0]
+                if base_type != self.sources[source_id]["type"]["type"]:
+                    return test.FAIL("Source {} state does not match the base type".format(source_id))
+        except KeyError:
+            return test.FAIL("Source {} JSON data is invalid".format(source_id))
+
+        return test.PASS()
+
+    def test_04_01(self, test):
+        """Each number Source type describes a valid set of allowed values"""
+        self.do_collect_sources(test)
+
+        found_number = False
+
+        try:
+            for source_id in self.sources:
+                source = self.sources[source_id]
+                if "number" != source["type"]["type"]:
+                    continue
+                # all 'enum' types define the allowed values
+                if "values" in source["type"]:
+                    continue
+                found_number = True
+
+                min = self.get_number(source["type"]["min"])

```

Comment:
In my opinion it would be useful for test_04_01 to check if the **min** and **max** keys are present and return a more specific error if they are not.

[Forked repo with a suggestion as to how it might look like](https://github.com/cristian-recoseanu/nmos-testing/commit/92f4a0a9a70004331be57e1846e15ee0fafd55a1)

Reply:
Thanks, @christian-recoseanu. We generally try to avoid re-testing things that are caught by the schema validation. If the 'min' or 'max' keys were missing, one of the auto tests should fail, with the message "Response schema validation error". I've just tried it and it's `auto_events_7`. Agreed, this message isn't as detailed as it could be, but it's clear how to investigate.

Reply:
We realised that the auto tests only try one 'child resource' which may of course be one that validates OK. So I've pushed the change to improve the message to indicate the missing key.